fw/http_parser.c

```diff
@@ -8600,45 +8699,140 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
 			&parser->hdr, {
-			parser->_acc = TFW_HTTP_CC_PROXY_REVAL;
+			parser->cache_control.dir_flag = TFW_HTTP_CC_PROXY_REVAL;
 		}, Resp_I_CC_p, Resp_I_Flag);
 		TRY_STR_INIT();
 		__FSM_I_JMP(Resp_I_Ext);
 	}
 
+	/*
+	 * no-cache or private with arguments as defined by rfc7234
+	 * sections 5.2.2.2 and 5.2.2.6
+	 */
+	__FSM_STATE(Resp_I_Flag_No_Cache) {
+		if (c == '=') {
+			/* Qualified form cancels out the general flag */
+			/*
+			 * Resp_I_CC_Opening_Quote => Resp_I_CC_Dir_Arg_Token =>
+			 * => Resp_I_CC_Dir_Arg_EoT
+			 */
+			__FSM_I_MOVE_fixup(Resp_I_CC_Dir_Opening_Quote, 1, 0);
+		}
+		__FSM_I_JMP(Resp_I_Flag);
+	}
+
+	__FSM_STATE(Resp_I_Flag_Private) {
+		if (c == '=')
+			__FSM_I_MOVE_fixup(Resp_I_CC_Dir_Opening_Quote, 1, 0);
+
+		__FSM_I_JMP(Resp_I_Flag);
+	}
+
 	__FSM_STATE(Resp_I_Flag) {
-		WARN_ON_ONCE(!parser->_acc);
+		/* A start of a standard directive successfully detected */
 		if (IS_WS(c) || c == ',' || IS_CRLF(c)) {
-			resp->cache_ctl.flags |= parser->_acc;
-			parser->_acc = 0;
+			resp->cache_ctl.flags |= parser->cache_control.dir_flag;
 			__FSM_I_JMP(Resp_I_EoT);
 		}
+		/* ...but the flag appears to have an unknown suffix */
 		__FSM_I_JMP(Resp_I_Ext);
 	}
 
+	__FSM_STATE(Resp_I_CC_Dir_Opening_Quote) {
+		if (c != '"')
+			__FSM_EXIT(TFW_BLOCK);
+
+		__FSM_I_MOVE_fixup(Resp_I_CC_Dir_Arg_Token, 1, 0);
+	}
+
+	__FSM_STATE(Resp_I_CC_Dir_Arg_Token) {
+		/* comma-separated field list (field-name/token) with optional
+		 * linear white space
+		 */
+		TfwStr *tokens;
+		tokens = parser->cache_control.dir_flag == TFW_HTTP_CC_PRIVATE ?
+			 &resp->private_tokens : &resp->no_cache_tokens;
+
+		__FSM_I_MATCH_MOVE_fixup_finish(token, Resp_I_Ext, 0, {
+			if (tfw_str_array_append_chunk(msg->pool, tokens,
+						       p, __fsm_sz,
+						       NULL, 0))
+			{
+				__FSM_EXIT(TFW_BLOCK);
+			}
+		});
+
+		/* Header name comparison requires a trailing colon */
+		if (tfw_str_array_append_chunk(msg->pool, tokens,
+					       p, __fsm_sz,
+					       ":", 1))
+		{
+			__FSM_EXIT(TFW_BLOCK);
+		}
+
+		__FSM_I_MOVE_fixup(Resp_I_CC_Dir_Arg_EoT, __fsm_sz, 0);
+	}
+
+	__FSM_STATE(Resp_I_CC_Dir_Arg_EoT) {
+		if (IS_WS(c))
+			__FSM_I_MOVE_fixup(Resp_I_CC_Dir_Arg_EoT, 1,
+					   TFW_STR_OWS);
+		if (IS_CRLF(c))
+			__FSM_EXIT(TFW_BLOCK);
+		 if (c == ',')
+			__FSM_I_MOVE_fixup(Resp_I_CC_Dir_Arg_Comma, 1, 0);
+		if (c == '"')
+			__FSM_I_MOVE_fixup(Resp_I_EoT, 1, 0);
+		/*
+		 * if (IS_TOKEN(c)) then block too, because tokens should be
+		 * separated by commas
+		 */
+		__FSM_EXIT(TFW_BLOCK);
+	}
+
+	__FSM_STATE(Resp_I_CC_Dir_Arg_Comma) {
+		if (IS_WS(c))
+			__FSM_I_MOVE_fixup(Resp_I_CC_Dir_Arg_Comma, 1,
+					   TFW_STR_OWS);
+		if (IS_CRLF(c))
+			__FSM_EXIT(TFW_BLOCK);
+		/*
+		 * Two consecutive commas in arguments not allowed,
+		 * just like in Cache-Control directives.
+		 */
+		if (c == ',')
+			__FSM_EXIT(TFW_BLOCK);
+		if (c == '"')
+			__FSM_I_MOVE_fixup(Resp_I_EoT, 1, 0);
```

Comment:
This state is to allow `no-cache="Set-Cookie, "` - is this intentional? If yes, then comment is required. Otherwise, we can just omit the check and block `"` on the default transition to `TFW_BLOCK`